From 8605eccff7a3870593d9807c8c624f89f7f71ff7 Mon Sep 17 00:00:00 2001
From: Victor Gu <xigu@marvell.com>
Date: Mon, 11 Apr 2016 17:27:09 +0800
Subject: [PATCH 1161/1240] clk: a3700: add API to get CPU clock source and
 divider

This patch adds one API 'get_cpu_clk_src_div' to
get current CPU clock TBG source selection,
and the prescaling divider number.

Change-Id: I134950a951ba94dc5f2f4e8f28884e2e5ef06eb6
Signed-off-by: Victor Gu <xigu@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/28907
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Reviewed-by: Kostya Porotchkin <kostap@marvell.com>
---
 arch/arm/cpu/armv8/armada3700/mvebu_clock.c  | 25 +++++++++++++++++++++++++
 arch/arm/include/asm/arch-armada3700/clock.h | 12 ++++++++++++
 2 files changed, 37 insertions(+)

diff --git a/arch/arm/cpu/armv8/armada3700/mvebu_clock.c b/arch/arm/cpu/armv8/armada3700/mvebu_clock.c
index f0c72dc..de71900 100644
--- a/arch/arm/cpu/armv8/armada3700/mvebu_clock.c
+++ b/arch/arm/cpu/armv8/armada3700/mvebu_clock.c
@@ -601,3 +601,28 @@ u32 get_ddr_clk(void)
 
 	return tbg>>1;
 }
+
+/******************************************************************************
+* Name: get_cpu_clk_src_div
+*
+* Description: Get CPU clock source selection and prescaling divider
+*
+* Input:	None
+* Output:	cpu_clk_sel: CPU clock source selection
+*		cpu_clk_prscl: CPU clock prescaling divider
+* Return:	Non-zero if failed to get the CPU clock selection and prescaling
+******************************************************************************/
+int get_cpu_clk_src_div(u32 *cpu_clk_sel, u32 *cpu_clk_prscl)
+{
+	/* 1. check cpu clock select */
+	if (!((readl(MVEBU_NORTH_CLOCK_SELECT_REG) >> 15) & 0x1))
+		return -1; /* CPU clock is using XTAL output*/
+
+	/* 2. get TBG select */
+	*cpu_clk_sel = (readl(MVEBU_NORTH_CLOCK_TBG_SELECT_REG) >> 22) & 0x3;
+
+	/* 3. get CPU clk divider */
+	*cpu_clk_prscl = (readl(MVEBU_NORTH_CLOCK_DIVIDER_SELECT0_REG) >> 28) & 0x7;
+
+	return 0;
+}
diff --git a/arch/arm/include/asm/arch-armada3700/clock.h b/arch/arm/include/asm/arch-armada3700/clock.h
index 9a6bd2e..f68e26a 100644
--- a/arch/arm/include/asm/arch-armada3700/clock.h
+++ b/arch/arm/include/asm/arch-armada3700/clock.h
@@ -577,6 +577,18 @@ u32 get_ddr_clk(void);
 ******************************************************************************/
 u32 set_clocks(u32 cpu_clk_mhz, u32 ddr_clk_mhz, u32 tbg_a_kvco_mhz, u32 tbg_b_kvco_mhz);
 
+/******************************************************************************
+* Name: get_cpu_clk_src_div
+*
+* Description: Get CPU clock source selection and prescaling divider
+*
+* Input:	None
+* Output:	cpu_clk_sel: CPU clock source selection
+*		cpu_clk_prscl: CPU clock prescaling divider
+* Return:	Non-zero if failed to get the CPU clock selection and prescaling
+******************************************************************************/
+int get_cpu_clk_src_div(u32 *cpu_clk_sel, u32 *cpu_clk_prscl);
+
 int init_clock(void);
 
 #endif /* _ARMADA3700_CLOCK_H_ */
-- 
1.9.1

